<a href="https://colab.research.google.com/github/asadbek11211/Pandas_exercises/blob/main/need_functions.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Foydali funksiyalar

No-odatiy qiymatlarni topish

In [ ]:
import pandas as pd
import numpy as np

In [ ]:
df = pd.read_csv("https://github.com/anvarnarz/praktikum_datasets/raw/main/MOCK_USER_DATA.csv", index_col=0)
df.head()

first_name   last_name                           email       gender   age
id                                                                          
1     Dorothy     Verryan               dverryan0@who.int   Non-binary     1
2     Pietrek     Scanlin          pscanlin1@yolasite.com         Male    91
3      Lilith    Sarjeant  lsarjeant2@creativecommons.org  Genderfluid  1233
4        Knox  Cunningham    kcunningham3@yellowpages.com  Genderfluid     6
5    Mohammed  Habbershon    mhabbershon4@sourceforge.net       Female    48

In [ ]:
df.describe()

age
count   100.000000
mean    145.660000
std     565.631411
min     -85.000000
25%      22.750000
50%      53.500000
75%      84.250000
max    4444.000000

Nimadur xato: `max` yosh `4444` deb ko'rsatayapti, tabiiyku bunday bo'lishi mumkin emas. `min` da ham manfiy qiymat ko'rishimiz mumkin, bu ham xato. Demak jadvalimizda noto'g'ri qiymatlar ham kirib qolgan. 

Keling, boshlanishiga katta qiymatlarni filtrlab olamiz, yaqin tarixda odam yoshi maksimum `123` dan oshmagnini bilgan holda, keling shu qiymatdan katta qatorlarni topamiz:

In [ ]:
df[df.age > 123]

first_name  last_name                           email       gender   age
id                                                                          
3       Lilith   Sarjeant  lsarjeant2@creativecommons.org  Genderfluid  1233
39      Godard      Miere               gmiere12@1688.com   Polygender  3434
47       Cleve  Sparshatt      csparshatt1a@economist.com  Genderfluid   333
49       Trace   Stockney          tstockney1c@flavors.me       Female   344
65      Eloise    Snoddon    esnoddon1s@elegantthemes.com  Genderfluid   332
80       Gabie     Rothon              grothon27@ehow.com  Genderqueer  4444
100      Mable    Toffano          mtoffano2r@myspace.com     Bigender   233

In [ ]:
df[df.age < 0]

first_name  last_name                           email       gender  age
id                                                                        
11     Elijah      Heine                eheinea@imdb.com     Bigender  -85
19      Luisa  Bradborne            lbradbornei@fema.gov       Female  -75
21      Levin     Worner                lwornerk@ihg.com  Genderqueer  -43
33   Ebenezer    Greeves           egreevesw@myspace.com         Male  -26
38       Hana   Langwade  hlangwade11@washingtonpost.com  Genderqueer  -42
90  Zachariah      Meins            zmeins2h@weather.com         Male  -69

no-odatiy qiymatlarni ajratib olamiz

In [ ]:
df[(df.age > 123) | (df.age < 0)]

first_name  last_name                           email       gender   age
id                                                                          
3       Lilith   Sarjeant  lsarjeant2@creativecommons.org  Genderfluid  1233
11      Elijah      Heine                eheinea@imdb.com     Bigender   -85
19       Luisa  Bradborne            lbradbornei@fema.gov       Female   -75
21       Levin     Worner                lwornerk@ihg.com  Genderqueer   -43
33    Ebenezer    Greeves           egreevesw@myspace.com         Male   -26
38        Hana   Langwade  hlangwade11@washingtonpost.com  Genderqueer   -42
39      Godard      Miere               gmiere12@1688.com   Polygender  3434
47       Cleve  Sparshatt      csparshatt1a@economist.com  Genderfluid   333
49       Trace   Stockney          tstockney1c@flavors.me       Female   344
65      Eloise    Snoddon    esnoddon1s@elegantthemes.com  Genderfluid   332
80       Gabie     Rothon              grothon27@ehow.com  Genderqueer  4444
90   Zachariah      Meins            zmeins2h@weather.com         Male   -69
100      Mable    Toffano          mtoffano2r@myspace.com     Bigender   233

In [ ]:
df['age'] = df.age.apply(np.abs)
df.describe()

age
count   100.000000
mean    152.460000
std     563.818411
min       1.000000
25%      32.000000
50%      59.000000
75%      85.000000
max    4444.000000

In [ ]:
ind = df[df.age > 123].index
df = df.drop(index=ind)

In [ ]:
df.describe()

age
count   93.000000
mean    52.612903
std     29.978265
min      1.000000
25%     29.000000
50%     52.000000
75%     77.000000
max    100.000000

In [ ]:
df

first_name   last_name                           email       gender  age
id                                                                         
1     Dorothy     Verryan               dverryan0@who.int   Non-binary    1
2     Pietrek     Scanlin          pscanlin1@yolasite.com         Male   91
4        Knox  Cunningham    kcunningham3@yellowpages.com  Genderfluid    6
5    Mohammed  Habbershon    mhabbershon4@sourceforge.net       Female   48
6       Glynn      Ebbing          gebbing5@rakuten.co.jp  Genderfluid   77
..        ...         ...                             ...          ...  ...
95      Byron  Whittleton   bwhittleton2m@theatlantic.com   Non-binary    2
96       Baxy       Marcu        bmarcu2n@list-manage.com      Agender   64
97     Kattie     Claxson          kclaxson2o@booking.com       Female   18
98       Clem     Fullerd  cfullerd2p@creativecommons.org     Bigender   55
99  Desdemona     Jurasek        djurasek2q@aboutads.info  Genderfluid   46

[93 rows x 5 columns]

Ko'rib turganingizdek, jadval qiymatlari normal holatga qaytdi.

Jadvlada yana bir e'tiborimni tortgna jihati, jinslar faqatgina erkak (`Male`) va ayol (`Female`) emas. 

In [ ]:
df.gender.unique()

array(['Non-binary', 'Male', 'Genderfluid', 'Female', 'Agender',
       'Bigender', 'Polygender', 'Genderqueer'], dtype=object)

In [ ]:
df2 = df[(df.gender == 'Male') | (df.gender == 'Female')]

In [ ]:
df2

first_name     last_name                         email  gender  age
id                                                                    
2     Pietrek       Scanlin        pscanlin1@yolasite.com    Male   91
5    Mohammed    Habbershon  mhabbershon4@sourceforge.net  Female   48
10      David      Meredyth    dmeredyth9@telegraph.co.uk  Female   68
12  Raffaello     Firminger     rfirmingerb@bloomberg.com  Female   74
17      Byram   Martynikhin     bmartynikhing@jiathis.com    Male   40
19      Luisa     Bradborne          lbradbornei@fema.gov  Female   75
22        Ray         Padly           rpadlyl@twitpic.com    Male   85
26   Ephrayim    Westmacott         ewestmacottp@ning.com    Male   77
29      Biron      Acklands        backlandss@comsenz.com    Male   13
33   Ebenezer       Greeves         egreevesw@myspace.com    Male   26
34     Brooks          Vose             bvosex@amazon.com  Female   34
36        Axe       Arundel           aarundelz@yahoo.com  Female   14
43       Bili     Swetenham        bswetenham16@europa.eu    Male   23
46     Alvira  Stollsteiner     astollsteiner19@mysql.com    Male   66
58     Brynna        Worsom              bworsom1l@wp.com    Male   66
61       Page       Merrell     pmerrell1o@slideshare.net    Male    1
62     Kennan         Visco          kvisco1p@jiathis.com  Female   94
63   Adrianne       Gommowe  agommowe1q@deliciousdays.com    Male  100
67     Bunnie        Ozanne      bozanne1u@whitehouse.gov    Male   92
70      Ambur        Aickin           aaickin1x@about.com    Male   60
72   Eleanora       Kamenar       ekamenar1z@examiner.com    Male   47
74      Curry    Smalecombe  csmalecombe21@washington.edu    Male   69
76      Misty         Blanc         mblanc23@amazon.co.jp    Male   74
77     Rudolf        Kalewe           rkalewe24@fotki.com    Male   12
81       Alex         Darth    adarth28@ezinearticles.com  Female   98
86    Rogerio          Akid           rakid2d@latimes.com  Female    3
87      Peder        Murrie       pmurrie2e@bloglovin.com    Male   69
88       Kain     Dumphries  kdumphries2f@hugedomains.com    Male   93
90  Zachariah         Meins          zmeins2h@weather.com    Male   69
91      Judas        Maffin         jmaffin2i@dropbox.com    Male    3
94     Sheree        Cockin     scockin2l@accuweather.com  Female   94
97     Kattie       Claxson        kclaxson2o@booking.com  Female   18

In [ ]:
df3 = df[df.gender.isin(['Male','Female'])]
df3

first_name     last_name                         email  gender  age
id                                                                    
2     Pietrek       Scanlin        pscanlin1@yolasite.com    Male   91
5    Mohammed    Habbershon  mhabbershon4@sourceforge.net  Female   48
10      David      Meredyth    dmeredyth9@telegraph.co.uk  Female   68
12  Raffaello     Firminger     rfirmingerb@bloomberg.com  Female   74
17      Byram   Martynikhin     bmartynikhing@jiathis.com    Male   40
19      Luisa     Bradborne          lbradbornei@fema.gov  Female   75
22        Ray         Padly           rpadlyl@twitpic.com    Male   85
26   Ephrayim    Westmacott         ewestmacottp@ning.com    Male   77
29      Biron      Acklands        backlandss@comsenz.com    Male   13
33   Ebenezer       Greeves         egreevesw@myspace.com    Male   26
34     Brooks          Vose             bvosex@amazon.com  Female   34
36        Axe       Arundel           aarundelz@yahoo.com  Female   14
43       Bili     Swetenham        bswetenham16@europa.eu    Male   23
46     Alvira  Stollsteiner     astollsteiner19@mysql.com    Male   66
58     Brynna        Worsom              bworsom1l@wp.com    Male   66
61       Page       Merrell     pmerrell1o@slideshare.net    Male    1
62     Kennan         Visco          kvisco1p@jiathis.com  Female   94
63   Adrianne       Gommowe  agommowe1q@deliciousdays.com    Male  100
67     Bunnie        Ozanne      bozanne1u@whitehouse.gov    Male   92
70      Ambur        Aickin           aaickin1x@about.com    Male   60
72   Eleanora       Kamenar       ekamenar1z@examiner.com    Male   47
74      Curry    Smalecombe  csmalecombe21@washington.edu    Male   69
76      Misty         Blanc         mblanc23@amazon.co.jp    Male   74
77     Rudolf        Kalewe           rkalewe24@fotki.com    Male   12
81       Alex         Darth    adarth28@ezinearticles.com  Female   98
86    Rogerio          Akid           rakid2d@latimes.com  Female    3
87      Peder        Murrie       pmurrie2e@bloglovin.com    Male   69
88       Kain     Dumphries  kdumphries2f@hugedomains.com    Male   93
90  Zachariah         Meins          zmeins2h@weather.com    Male   69
91      Judas        Maffin         jmaffin2i@dropbox.com    Male    3
94     Sheree        Cockin     scockin2l@accuweather.com  Female   94
97     Kattie       Claxson        kclaxson2o@booking.com  Female   18

### Jadvlani aralashtirib tashlash

Ba'zida, tadqiqod uchun jadvaldagi qiymatlar o'rnini aralashtirib tashlash talab qilinishi mumkin. Misol uchun, sizda ma'lumotlar ma'lum bir tartibda turibdi, lekin siz ma'lumotlarni tasodifiy tartibda chiqarmoqchiisiz, buning uchun `numpy` kutubxonasining `random` modulidagi `permutation` funksiyasiga murojaat qilamiz.

Bizda quyidagi jadval bor:

In [ ]:
df3.head()

first_name    last_name                         email  gender  age
id                                                                   
2     Pietrek      Scanlin        pscanlin1@yolasite.com    Male   91
5    Mohammed   Habbershon  mhabbershon4@sourceforge.net  Female   48
10      David     Meredyth    dmeredyth9@telegraph.co.uk  Female   68
12  Raffaello    Firminger     rfirmingerb@bloomberg.com  Female   74
17      Byram  Martynikhin     bmartynikhing@jiathis.com    Male   40

In [ ]:
len(df3)

32

In [ ]:
df3.shape

(32, 5)

In [ ]:
df3.size

160

In [ ]:
list = np.random.permutation(32)
list

array([ 5, 21, 19, 10, 26, 31,  1,  0, 23, 13,  9,  6, 27,  8, 28,  2, 30,
       25,  4, 20, 14, 16,  3, 17, 12, 15, 24, 29, 18,  7, 11, 22])

In [ ]:
df3.index = list

In [ ]:
df3.head(10).sort_index()

first_name    last_name                         email  gender  age
0    Ephrayim   Westmacott         ewestmacottp@ning.com    Male   77
1         Ray        Padly           rpadlyl@twitpic.com    Male   85
5     Pietrek      Scanlin        pscanlin1@yolasite.com    Male   91
10  Raffaello    Firminger     rfirmingerb@bloomberg.com  Female   74
13   Ebenezer      Greeves         egreevesw@myspace.com    Male   26
19      David     Meredyth    dmeredyth9@telegraph.co.uk  Female   68
21   Mohammed   Habbershon  mhabbershon4@sourceforge.net  Female   48
23      Biron     Acklands        backlandss@comsenz.com    Male   13
26      Byram  Martynikhin     bmartynikhing@jiathis.com    Male   40
31      Luisa    Bradborne          lbradbornei@fema.gov  Female   75

In [ ]:
df3.take(list).head(10).sort_index()

first_name     last_name                         email  gender  age
5     Pietrek       Scanlin        pscanlin1@yolasite.com    Male   91
8      Alvira  Stollsteiner     astollsteiner19@mysql.com    Male   66
9      Brooks          Vose             bvosex@amazon.com  Female   34
16      Curry    Smalecombe  csmalecombe21@washington.edu    Male   69
17     Rudolf        Kalewe           rkalewe24@fotki.com    Male   12
20      Ambur        Aickin           aaickin1x@about.com    Male   60
21   Mohammed    Habbershon  mhabbershon4@sourceforge.net  Female   48
22     Kattie       Claxson        kclaxson2o@booking.com  Female   18
24      Peder        Murrie       pmurrie2e@bloglovin.com    Male   69
31      Luisa     Bradborne          lbradbornei@fema.gov  Female   75